File: packages/oxlint-plugin-react-doctor/src/plugin/rules/security-scan/jwt-insecure-verification.ts
import { defineRule } from "../../utils/define-rule.js";
import type { ScanFinding } from "../../utils/file-scan.js";
import { getLocationAtIndex } from "./utils/get-location-at-index.js";
import { isProductionSourcePath } from "./utils/is-production-source-path.js";
import { getScannableContent } from "./utils/scan-by-pattern.js";

// `alg: none` / `algorithms: ["none"]` disables signature verification — any
// attacker-forged token is accepted. This is the one unambiguous,
// text-detectable JWT misconfiguration. Detecting an *unpinned* `jwt.verify`
// (missing `algorithms` allowlist → RS256→HS256 confusion) precisely requires
// scope-aware resolution of the options argument/binding (variable vs inline
// vs callback, import scope, TS annotations), which a regex scan cannot do
// without false positives/negatives — that is deferred to a future AST rule.
const NONE_ALGORITHM_PATTERN = /\b(?:alg|algorithms?)\s*:\s*\[?\s*["'`]none["'`]/gi;

// True when `index` falls inside a string/template literal — used to skip an
// `algorithm: "none"` mentioned in error text or docs (`"never use algorithm:
// 'none'"`) rather than a real options object. Template `${…}` interpolations
// are treated as code, so a real `alg: "none"` inside one is NOT skipped.
const isIndexInsideStringLiteral = (content: string, index: number): boolean => {
  let stringDelimiter: string | null = null;
  // Brace depth of each open `${…}` expression nested inside template strings.
  const templateExpressionDepths: number[] = [];
  for (let cursor = 0; cursor < index; cursor += 1) {
    const character = content[cursor];
    if (stringDelimiter === "`") {
      if (character === "\\") {
        cursor += 1;
      } else if (character === "`") {
        stringDelimiter = null;
      } else if (character === "$" && content[cursor + 1] === "{") {
        templateExpressionDepths.push(0);
        stringDelimiter = null;
        cursor += 1;
      }
      continue;
    }
    if (stringDelimiter !== null) {
      if (character === "\\") {
        cursor += 1;
      } else if (character === stringDelimiter) {
        stringDelimiter = null;
      }
      continue;
    }
    if (character === '"' || character === "'" || character === "`") {
      stringDelimiter = character;
    } else if (templateExpressionDepths.length > 0) {
      const top = templateExpressionDepths.length - 1;
      if (character === "{") {
        templateExpressionDepths[top] += 1;
      } else if (character === "}") {
        if (templateExpressionDepths[top] === 0) {
          templateExpressionDepths.pop();
          stringDelimiter = "`";
        } else {
          templateExpressionDepths[top] -= 1;
        }
      }
    }
  }
  return stringDelimiter !== null;
};

export const jwtInsecureVerification = defineRule({
  id: "jwt-insecure-verification",
  title: "JWT verified with the 'none' algorithm",
  severity: "error",
  recommendation:
    "Never accept the `none` algorithm; it disables signature verification and lets any forged token through. Pin the real algorithm(s) explicitly (`jwt.verify(token, key, { algorithms: ['RS256'] })`).",
  scan: (file) => {
    if (!isProductionSourcePath(file.relativePath)) return [];
    const content = getScannableContent(file);
    if (!/\bjwt\b|jsonwebtoken|\bjose\b/i.test(content)) return [];

    const findings: ScanFinding[] = [];
    NONE_ALGORITHM_PATTERN.lastIndex = 0;
    for (
      let noneMatch = NONE_ALGORITHM_PATTERN.exec(content);
      noneMatch !== null;
      noneMatch = NONE_ALGORITHM_PATTERN.exec(content)
    ) {
      if (isIndexInsideStringLiteral(content, noneMatch.index)) continue;
      const location = getLocationAtIndex(content, noneMatch.index);
      findings.push({
        message:
          "JWT is configured with the 'none' algorithm, which disables signature verification, so any forged token is accepted.",
        line: location.line,
        column: location.column,
      });
    }
    return findings;
  },
});

File: native/oxlint/scans/jwt-insecure-verification.rs
use lazy_regex::{Lazy, Regex, lazy_regex};

use super::{ScanFinding, get_location_at_index::get_location_at_index};

const MESSAGE: &str = "JWT is configured with the 'none' algorithm, which disables signature verification, so any forged token is accepted.";

static JWT_EVIDENCE_PATTERN: Lazy<Regex> = lazy_regex!(r"(?i)\bjwt\b|jsonwebtoken|\bjose\b");
static NONE_ALGORITHM_PATTERN: Lazy<Regex> =
    lazy_regex!(r#"(?i)\b(?:alg|algorithms?)\s*:\s*\[?\s*["'`]none["'`]"#);

pub fn scan(relative_path: &str, source: &str) -> Vec<ScanFinding> {
    if !super::is_production_file_path::is_production_source_path(relative_path) {
        return Vec::new();
    }
    let scannable =
        super::get_scannable_content::get_scannable_content(relative_path, source, false);
    let content = super::normalize_js_regex_content::normalize_js_regex_content(&scannable);
    if !JWT_EVIDENCE_PATTERN.is_match(&content) {
        return Vec::new();
    }
    let findings = NONE_ALGORITHM_PATTERN
        .find_iter(&content)
        .filter(|found| !is_inside_string_literal(&content, found.start()))
        .map(|found| {
            let (line, column) = get_location_at_index(source, &content, found.start());
            ScanFinding::inherited(MESSAGE, line, column)
        })
        .collect::<Vec<_>>();
    findings
}

fn is_inside_string_literal(content: &str, index: usize) -> bool {
    let mut delimiter = None;
    let mut template_expression_depths = Vec::<usize>::new();
    let mut cursor = 0;
    while cursor < index {
        let byte = content.as_bytes()[cursor];
        if delimiter == Some(b'`') {
            if byte == b'\\' {
                cursor = escaped_character_end(content, cursor).min(index);
            } else if byte == b'`' {
                delimiter = None;
                cursor += 1;
            } else if byte == b'$' && content.as_bytes().get(cursor + 1) == Some(&b'{') {
                template_expression_depths.push(0);
                delimiter = None;
                cursor += 2;
            } else {
                cursor += content[cursor..].chars().next().map_or(1, char::len_utf8);
            }
            continue;
        }
        if let Some(active_delimiter) = delimiter {
            if byte == b'\\' {
                cursor = escaped_character_end(content, cursor).min(index);
            } else {
                if byte == active_delimiter {
                    delimiter = None;
                }
                cursor += content[cursor..].chars().next().map_or(1, char::len_utf8);
            }
            continue;
        }
        if matches!(byte, b'\'' | b'"' | b'`') {
            delimiter = Some(byte);
        } else if let Some(depth) = template_expression_depths.last_mut() {
            if byte == b'{' {
                *depth += 1;
            } else if byte == b'}' {
                if *depth == 0 {
                    template_expression_depths.pop();
                    delimiter = Some(b'`');
                } else {
                    *depth -= 1;
                }
            }
        }
        cursor += content[cursor..].chars().next().map_or(1, char::len_utf8);
    }
    delimiter.is_some()
}

fn escaped_character_end(source: &str, backslash_index: usize) -> usize {
    let escaped_start = backslash_index + 1;
    source
        .get(escaped_start..)
        .and_then(|tail| tail.chars().next())
        .map_or(source.len(), |character| {
            escaped_start + character.len_utf8()
        })
}
